design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1196 @@
+			return true
+		}
+	}
+	return false
+}
+
+// fetchResources fetches resources that match the given selector.
+func fetchResources(ctx context.Context, client client.Reader, selector *v1beta1.ResourceSelector) ([]*unstructured.Unstructured, error) {
+	// Fetch resources by name or label selector...
+	// return nil on not found error
+}
+
+// Check if the references are in the observed state of the request.
+func checkAllowedWatchedResources(alreadyRequestedResources map[string]bool, refs map[string]v1beta1.ResourceRef) error {
+	for name, ref := range refs {
+		if !alreadyRequestedResources[name] {
+            return errors.Errorf( /* ... */ )
+        }
+	}
+	return nil
+}
+```
+
+With the above proposal what currently can be implemented as follows:
+```yaml
+---
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: nop.sqlinstances.example.org
+spec:
+  compositeTypeRef:
+    apiVersion: example.org/v1alpha1
+    kind: XSQLInstance
+  environment:
+    environmentConfigs:
+      - ref:
+          name: example-environment-1
+      - type: Selector
+        selector:
+          mode: Multiple
+          sortByFieldPath: data.priority
+          matchLabels:
+            - type: FromCompositeFieldPath
+              key: stage
+              valueFromFieldPath: metadata.labels[stage]
+  resources:
+    - name: nop
+      base:
+        apiVersion: nop.crossplane.io/v1alpha1
+        kind: NopResource
+        spec:
+          forProvider:
+            conditionAfter:
+              - conditionType: Ready
+                conditionStatus: "False"
+                time: 0s
+              - conditionType: Ready
+                conditionStatus: "True"
+                time: 1s
+      patches:
+        - type: FromEnvironmentFieldPath
+          fromFieldPath: complex.c.f
+          toFieldPath: metadata.annotations[valueFromEnv]
+```
+
+Could be rewritten as follows, assuming we implemented a Function (e.g.
+`function-environment-configs` or a more generic `function-extra-resources`)
+that can request `EnvironmentConfigs` as extra resources, merge them according
+to its logic and set the result as in the `Context`, at the key
+`function-patch-and-transform` conventionally expects it,
+`apiextensions.crossplane.io/environment`:
+```yaml
+---
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: nop.sqlinstances.example.org
+spec:
+  compositeTypeRef:
+    apiVersion: example.org/v1alpha1
+    kind: XSQLInstance
+  mode: Pipeline
+  pipeline:
+  - name: envconfigs
+    functionRef:
+      name: function-environment-configs
+    allowedExtraResources:
+    - apiVersion: apiextensions.crossplane.io/v1alpha1
+      kind: EnvironmentConfig
+    input:
+      apiVersion: extra.fn.crossplane.io/v1beta1
+      kind: ExtraResourceSelectors
+      extraResourceSelectors:
+        environmentConfigs:
```

Comment:
I can imaging environmentConfig selection rendering just by looking at it:) It will be next level DX when working with EnvironmentConfigs 👍